core/templates/dev/head/pages/exploration-editor-page/services/exploration-rights-data.service.ts

```diff
@@ -0,0 +1,53 @@
+// Copyright 2019 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview A data service that stores data
+ * about the rights for this exploration.
+ * This is different from the exploration rights service
+ * since the data used here is required globally and may
+ * create a circular dependency with exploration data service
+ * which is used in exploration rights service
+ */
+require('services/ContextService.ts');
+require('services/contextual/UrlService.ts');
+var oppia = require('AppInit.ts').module;
+
+oppia.factory('ExplorationRightsDataService', [
```

Comment:
I'm fine with the current name, actually. I think "exploration rights" is a known object and we are consistent about its use in the backend, so I interpret this as a service which manages the "'exploration rights' data".

Reply:
Also, btw, please assign me if you need me to reply to something (and in general, assign anyone you "@"-mention so they know the ball is in their court).

Reply:
> please assign me if you need me to reply to something (and in general, assign anyone you "@"-mention so they know the ball is in their court).

@seanlip Sure, I'll do that.

Can you see [this](https://github.com/oppia/oppia/pull/7194/files#r307305949) comment? We already have a service named `ExplorationRightsService` so will it be fine to create `ExplorationRightsDataService`?

Reply:
Oh, that's interesting -- yes, I agree with you. Those filenames are too similar, and it's confusing.

The first thing I always ask in something like this is -- what does each file actually do? The fileoverview of this file is actually the same as that of ExplorationRightsService, which muddies things even further. It seems that the two services are storing the same things, which is not a good sign either because it sets up two sources of truth.

Also I don't understand why the data here is "required globally", given that this is just an exploration-related service.

I do think some changes need to happen here -- perhaps unify the services, clarify the purposes of the different files, etc. @Jamesjay4199, perhaps you could clarify.

Reply:
@seanlip, @DubeySandeep I initially wanted to do it in the exploration rights service, but I noticed that the service depends on exploration data service and the data to be gotten from the service will be needed bu the exploration data service, so this will create a circular dependency. That is the only reason this was given a service of its own.

Reply:
But that's not a good enough reason, in itself, for having two services. You need to look at the system as a whole and think: if I were building this from scratch, how would I organize it? Then implement that (which might involve changing existing systems too).

Reply:
Okay.
I'll do just that.

Reply:
I think the data can be gotten from the exploration data service or a different service since exploration rights service is heavily dependent on exploration data service. Or in a different service entirely.

Reply:
You'll need to propose something concrete, such that each file has a clear definition and that there is no duplication conceptually or code-wise.

After you are clear on where you want to get to, you can propose a series of steps that gets you from the present state to the desired end state.

Feel free to make a doc explaining clearly your solutions the above two points, and @vojtechjelinek and I would be happy to take a look.

Reply:
@seanlip, here's the [doc](https://docs.google.com/document/d/15Al5myoQasSy4FkPFUioUZqMyPhzaE1tBOH1a2h_0j4/edit?usp=drivesdk).
PTAL.

Reply:
Hi @Jamesjay4199, I mostly don't understand the proposed solution in your doc. Perhaps you could provide more detail, for example:

- What happens to ExplorationDataService? Does it still exist in your proposed solution?
- What happens to ExplorationRightsService? Does it still exist in your proposed solution?
- What does the "import graph" for the services look like in your proposed solution?
- What is the name of the new service?

In general, give enough detail to explain what the "final solution" actually looks like, so that the reviewers can understand the diff between the "before" and "after" states of the codebase. Currently what you have is, at best, a sketch of a solution.

Reply:
I thought about it and I made a few important changes to the doc.
Okay, I've updated the doc.
You can take a look at it now.
Thanks.

Reply:
I took a look. I don't think it makes sense for user rights data to be bundled with the ExplorationDataService. That muddies the definition of the service.

If the only thing that's needed is the exploration ID and version, it seems like that's something the client knows about. How about having both the ExplorationDataService and ExplorationRightsService depend on an ExplorationIdAndVersionService that stores and maintains a record of the ID and version of the currently-loaded exploration?

Reply:
Hi @seanlip, I tried doing this, but I noticed that the version is gotten after fetching the exploration.

Reply:
Hi @Jamesjay4199, I would need (much) more detail to address your (implied) question. Please explain what happened clearly when you tried doing this, so that an outsider can reconstruct and follow along with your steps.

(In general I think this pattern of "not explaining enough" keeps coming up; you might want to try giving clearer explanations at the outset in the future to avoid churn.)

Reply:
Okay. What I mean is:
The exploration version is gotten after fetching the exploration via ajax and fetching the exploration is done by the exploration data service

Reply:
@Jamesjay4199 Please be (much) more specific: include links to the relevant parts of the code in your explanation, and refer to specific services, function names and line numbers.

Reply:
@seanlip Originally, the job of the exploration data service is to fetch exploration and the version of the exploration can only be gotten after fetching the exploration. Therefore, having an explorationIdAndVersion service will need to fetch the exploration data (in order to get the version. The ID can be gotten from the URL). And, I don't think it will be a good idea to have two services that fetch the exploration.

Reply:
@Jamesjay4199 I'll repeat my previous comment:

> Please be (much) more specific: **include links** to the relevant parts of the code in your explanation, and refer to **specific services, function names and line numbers**.

Do this for everything you reply to, otherwise it will take a lot of back and forth to figure out what the issue is -- we don't have the context that you already have in your mind.

> @seanlip, @DubeySandeep I initially wanted to do it in the exploration rights service, but I noticed that the service depends on exploration data service and the data to be gotten from the service will be needed by the exploration data service, so this will create a circular dependency. That is the only reason this was given a service of its own.

Can you rewrite this comment in a way that's more specific? E.g.: what is "it", what specific data is needed by the exploration data service (include a link to the line of code), what specific data produced by the rights service is needed by the exploration data service (include a link to the line of code), etc.? I started looking into this a bit to see if I could figure out a way to break the dependency and then realized I didn't know what the dependency actually was. Clearer, more precise explanations that refer to specific lines of code and specific constructs would help with this.

Reply:
> Can you rewrite this comment in a way that's more specific

I initially wanted to send the exploration rights data in the exploration rights service, but then I noticed that the exploration rights service [depends on the exploration data service](https://github.com/oppia/oppia/blob/aa6961882d83e3e863a4af049583c9b32255873e/core/templates/dev/head/pages/exploration-editor-page/services/exploration-rights.service.ts#L67). and this will lead to a circular dependency between exploration rights service and exploration data service

Reply:
So, I decided to use a new service for this purpose.  

>  So, how about having both the ExplorationDataService and ExplorationRightsService depend on an ExplorationIdAndVersionService that stores and maintains a record of the ID and version of the currently-loaded exploration

About the problem with this approach is that to get the version of the exploration, we will need to fetch the exploration via ajax since it is not available in the exploration editor URL.
The explorationDataService is saddled with the [responsibility of fetching the exploration via ajax](https://github.com/oppia/oppia/blob/aa6961882d83e3e863a4af049583c9b32255873e/core/templates/dev/head/pages/exploration-editor-page/services/exploration-data.service.ts#L19), so I don't think it would be a great idea for the explorationIDandVersion service to get the load the exploration again since what we want is the just the version (explorationID can be gotten from the exploration editor URL).

Reply:
> I initially wanted to send the exploration rights data in the exploration rights service, but then I noticed that the exploration rights service depends on the exploration data service. and this will lead to a circular dependency between exploration rights service and exploration data service

Where is the circular dependency? How does the exploration data service depend on the exploration rights service?

Reply:
> Where is the circular dependency? How does the exploration data service depend on the exploration rights service?

Currently, it does not. But the exploration data service needs [data like can_edit](https://github.com/oppia/oppia/blob/ac9483a7b664ae8a7d5407a16ef77e390d611e12/core/templates/dev/head/pages/exploration-editor-page/services/exploration-data.service.ts#L55), etc. and fetching these data from the exploration rights service will cause a circular dependency.

I think the best solution would be to fetch these data(can_edit, can_voiceover, etc.)  from a different service that wouldn't depend on any of these services. Which is what I did earlier, but the name of the service is where I would actually need help because I called it ExplorationRightsDataService which okay considering its similarity with ExplorationRightsService

Reply:
Ah, I see. This information is actually really helpful, because it changes my suggestion completely.

I think explorationData should be independent of stuff like can_edit / can_voiceover. I wonder why we need two different URLs for autosaving drafts, and why we can't have just one instead. I suggest chatting with @DubeySandeep about this to see whether that dependency can be removed.

Reply:
We resolved that we could make a `can_access_unpublished_exploration` acl and then do the role check in the service layer, then it would be possible to merge both urls.

However, I think that would be out of the scope of this PR.